devicetypes/stelpro/stelpro-ki-zigbee-thermostat.src/stelpro-ki-zigbee-thermostat.groovy

```diff
@@ -0,0 +1,576 @@
+def getATTRIBUTE_TEMP_DISP_MODE() { 0x0000 }
+def getATTRIBUTE_KEYPAD_LOCKOUT() { 0x0001 }
+
+
+def getSupportedThermostatModes() {
+	["heat", "eco", "off"]
+}
+
+def getMinSetpointIndex() {
+	0
+}
+def getMaxSetpointIndex() {
+	1
+}
+
+def getThermostatSetpointRange() {
+	(getTemperatureScale() == "C") ? [5, 30] : [41, 86]
+}
+
+def getHeatingSetpointRange() {
+	thermostatSetpointRange
+}
+
+def getSetpointStep() {
+	(getTemperatureScale() == "C") ? 0.5 : 1.0
+}
+
+def getModeMap() {[
+	"00":"off",
+	"04":"heat",
+	"05":"eco"
+]}
+
+def setupHealthCheck() {
+	// Device-Watch simply pings if no device events received for 32min(checkInterval)
+	sendEvent(name: "checkInterval", value: 2 * 15 * 60 + 2 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+}
+
+def configureSupportedRanges() {
+	sendEvent(name: "supportedThermostatModes", value: supportedThermostatModes, displayed: false)
+	sendEvent(name: "thermostatSetpointRange", value: thermostatSetpointRange, displayed: false)
+	sendEvent(name: "heatingSetpointRange", value: heatingSetpointRange, displayed: false)
+}
+
+def installed() {
+	sendEvent(name: "temperatureAlarm", value: "cleared", displayed: false)
+
+	setupHealthCheck()
+
+	configureSupportedRanges()
+}
+
+def updated() {
+	def requests = []
+	setupHealthCheck()
+
+	configureSupportedRanges()
+
+	unschedule(scheduledUpdateWeather)
+	if (settings.zipcode) {
+		requests += updateWeather()
+		runEvery1Hour(scheduledUpdateWeather)
+	}
+
+	requests += parameterSetting()
+	response(requests)
+}
+
+def parameterSetting() {
+	def lockmode = null
+	def valid_lock = false
+
+	log.debug "lock : $settings.lock"
+	if (settings.lock == "Yes") {
+		lockmode = 0x01
+		valid_lock = true
+	} else if (settings.lock == "No") {
+		lockmode = 0x00
+		valid_lock = true
+	}
+
+	if (valid_lock) {
+		log.debug "lock valid"
+		delayBetween([
+				zigbee.writeAttribute(THERMOSTAT_UI_CONFIG_CLUSTER, ATTRIBUTE_KEYPAD_LOCKOUT, DataType.ENUM8, lockmode),
+				poll()
+			], 200)
+	} else {
+		log.debug "nothing valid"
+	}
+}
+
+def parse(String description) {
+	log.debug "Parse description $description"
+	def map = [:]
+
+	// If the user installed with an old DTH version, update so that the new mobile client will work
+	if (!device.currentValue("supportedThermostatModes")) {
+		configureSupportedRanges()
+	}
+	// Existing installations need the temperatureAlarm state initialized
+	if (device.currentValue("temperatureAlarm") == null) {
+		sendEvent(name: "temperatureAlarm", value: "cleared", displayed: false)
+	}
+
+	if (description?.startsWith("read attr -")) {
+		def descMap = zigbee.parseDescriptionAsMap(description)
+		log.debug "Desc Map: $descMap"
+		if (descMap.clusterInt == THERMOSTAT_CLUSTER) {
+			if (descMap.attrInt == ATTRIBUTE_LOCAL_TEMP) {
+				map.name = "temperature"
+				map.unit = getTemperatureScale()
+				map.value = getTemperature(descMap.value)
+
+				if (descMap.value == "7ffd") {		// 0x7FFD
```

Comment:
Since these values have alpha characters it runs the risk of breaking if the case changes for some reason. It would be much better to convert these values to integers and then do numerical comparisons.

Reply:
Alright, just made the change to use strict numbers instead of strings.